cirq/sim/simulator.py

```diff
@@ -526,7 +525,52 @@ def _base_iterator(
         Yields:
             StepResults from simulating a Moment of the Circuit.
         """
-        raise NotImplementedError()
+        acton_args = self._create_act_on_args(circuit, initial_state, qubit_order)
+        return self._core_iterator(circuit, acton_args, qubit_order)
+
+    @abc.abstractmethod
+    def _create_act_on_args(
+        self,
+        circuit: circuits.Circuit,
+        initial_state: Any,
+        qubit_order: ops.QubitOrderOrList = ops.QubitOrder.DEFAULT,
+    ) -> TActOnArgs:
+        """Creates the ActOnArgs state for a simulator.
+
+        Args:
+            circuit: The circuit to simulate.
+            initial_state: The initial state for the simulation. The form of
+                this state depends on the simulation implementation. See
+                documentation of the implementing class for details.
+            qubit_order: Determines the canonical ordering of the qubits. This
+                is often used in specifying the initial state, i.e. the
+                ordering of the computational basis states.
+
+        Returns:
+            ActOnArgs for the simulator.
+        """
+
+    @abc.abstractmethod
+    def _core_iterator(
+        self,
+        circuit: circuits.Circuit,
+        initial_state: TActOnArgs,
+        qubit_order: ops.QubitOrderOrList = ops.QubitOrder.DEFAULT,
```

Comment:
Seems odd to require the qubit_order here if we've already passed it in above. Can this be stored somewhere?

Reply:
Plausibly if we stuff it into the ActOnArgs base class as a qubit_map, that would have the additional benefit that we'd no longer have to require simulators to manually set ActOnArgs.axes for each operation: the act_on protocol could look the axes up from the qubit_map and the operation qubits.

Reply:
Gave this a try but it ends up being a big change. Mostly the unit tests since all the ActOnArgs creations end up needing a new parameter. Maybe a separate PR would be better.